SPECIFICATION.md

```diff
@@ -0,0 +1,143 @@
+# The FontoBene Font Format, Version 1.0
+
+
+## Introduction
+
+FontoBene is a vector font format. It only uses strokes to form glyphs, in
+order for easy rendering in systems that cannot easily deal with more complex
+parts of formats like TTF or OTF, like Bézier curves.
+
+FontoBene was designed with the use case of text on printed circuit boards
+(PCBs) in mind. PCBs are usually manufactured using the Gerber format as an
+exchange format to describe the etching layers, which does not support
+quadratic / cubic curves.
+
+The main inspiration for this font format was the LFF font used by LibreCAD.
+
+
+## High Level Format Description
+
+Every FontoBene font file consists of UTF-8 encoded text. The lines are
+separated by newlines (U+000A LINE FEED). It starts with a header, followed by
+glyph definitions.
+
+
+### Header
+
+The header is parsed line-by-line, starting with a number sign (`#`) as the
+first character on every line. The number sign is followed by a space, a key,
+and a value. Key and value are separated by a colon (`:`). Everything following
+the colon is considered the value. Both keys and values are strings.
+
+Example:
+
+    # Format: FontoBene/1.0
+    # Creator: Max Müller
+    # Version: 1.0
+    # VersionDate: 2017-09-27
+    # License: GPLv3+
+
+All keys that may be used are specified in the specification. A user MAY use
+non-standard keys, but must prefix them with `X-`, for example `X-Foo: Bar`.
+
+Whitespace around key and value should be stripped by the parsers.
+
+
+### Body
+
+The first empty line separates the header from the body. The body consists of
+glyph definition blocks separated by at least one empty line.
+
+Every glyph definition starts with the glyph declaration, a unicode codepoint
+in square brackets (e.g. `[1F4A9]` for the pile of poo character). It may
+optionally be followed by a UTF-8 representation of the character for easier
+previewing (e.g. `💩`). However, that preview will be ignored by the parser.
```

Comment:
Maybe the "comment" after the character codepoint should be the Unicode name (e.g. "PILE OF POO") instead of the representation (e.g. "💩")? That would also work for glyphs not supported by editor / OS font, but it would make the font files larger.

Reply:
I would prefer the real representation because it keeps the files small... If an editor cannot display some of these characters, that's not really tragic IMO ;)

Reply:
Ok, let's keep this :)

Reply:
Another open question: Should we allow metadata for the glyphs? That would allow for things like kerning to be specified and give us some more flexibility in the future.

Since a polyline starts with a digit or with 0, and a reference starts with a capital C, we could simply allow lowercase key-value pairs for metadata anywhere in the glyph definition (but with the convention to put it immediately after the declaration).